suite/tests/client-interface/emulation_api_simple.dll.c

```diff
@@ -0,0 +1,165 @@
+ * DAMAGE.
+ *
+ */
+
+/* A simple AArch64 client replacing 'dst = src0 & src1' with
+ * 'dst = !(!src0 | !src1)' to sanity test two core emulation API functions:
+ * - drmgr_insert_emulation_start()
+ * - drmgr_insert_emulation_end()
+ *
+ * Note this emulation client is for AArch64 only:
+ * XXX i#3173 Improve testing of emulation API functions
+ */
+
+#include "dr_api.h"
+#include "drmgr.h"
+#include <limits.h>
+#include <stdint.h>
+#include <string.h>
+#include <unistd.h>
+
+/* Events handlers */
+static void
+event_exit(void);
+
+static dr_emit_flags_t
+event_instruction_change(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                         bool translating);
+
+DR_EXPORT void
+dr_client_main(client_id_t id, int argc, const char *argv[])
+{
+    if (!drmgr_init())
+        DR_ASSERT(false);
+
+    dr_register_exit_event(event_exit);
+
+    if (!drmgr_register_bb_app2app_event(event_instruction_change, NULL))
+        DR_ASSERT(false);
+}
+
+static void
+event_exit(void)
+{
+    if (!drmgr_unregister_bb_app2app_event(event_instruction_change))
+        DR_ASSERT(false);
+    drmgr_exit();
+}
+
+static dr_emit_flags_t
+event_instruction_change(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                         bool translating)
+{
+    instr_t *instr, *next_instr;
+
+    for (instr = instrlist_first_app(bb); instr != NULL; instr = next_instr) {
+        /* We're deleting some instrs, so get the next first. */
+        next_instr = instr_get_next_app(instr);
+
+        /* We replace 'dst = src0 & src1' with 'dst = !(!src0 | !src1)'. We only
+         * transform AND instructions with unshifted, 64 bit register operands. */
+        if (instr_get_opcode(instr) != OP_and)
+            continue;
+        opnd_t dst = instr_get_dst(instr, 0);
+        opnd_t src0 = instr_get_src(instr, 0);
+        opnd_t src1 = instr_get_src(instr, 1);
+
+        if (!opnd_is_reg(src0) || instr_num_srcs(instr) != 4 ||
+            opnd_get_size(src0) != OPSZ_8 ||
+            opnd_get_immed_int(instr_get_src(instr, 3)) != 0)
+            continue;
+
+        app_pc raw_instr_pc = instr_get_app_pc(instr);
+
+        /* Insert emulation label to signal start of emulation code sequence. The
+         * label is loaded with data about the instruction being emulated for use by
+         * an observational client.
+         */
+        emulated_instr_t emulated_instr;
+        emulated_instr.size = sizeof(emulated_instr);
+        emulated_instr.pc = raw_instr_pc;
+        emulated_instr.instr = instr;
+        drmgr_insert_emulation_start(drcontext, bb, instr, &emulated_instr);
+
+        dr_save_reg(drcontext, bb, instr, DR_REG_X26, SPILL_SLOT_1);
+        dr_save_reg(drcontext, bb, instr, DR_REG_X27, SPILL_SLOT_2);
+
+        opnd_t scratch0 = opnd_create_reg(DR_REG_X26);
+        opnd_t scratch1 = opnd_create_reg(DR_REG_X27);
+
+        /* scratch0 = !src0
+         * ORN <Xd>, <Xn>, <Xm>{, <shift> #<imm>}
```

Comment:
We check above that the shift is 0 and use shift = 0 all around here. I would get rid of the shift in the commend (or extend it to support shifts ;)); same below